authenticators/jwt_from_extra_provider.go

```diff
@@ -0,0 +1,85 @@
+package authenticators
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/arrikto/oidc-authservice/common"
+	goidc "github.com/coreos/go-oidc"
+)
+
+const bearerPrefix = "Bearer "
+
+type JWTFromExtraProviderAuthenticator struct {
+	headerName   string
+	issuer       string
+	isserName    string
+	remoteKeySet goidc.KeySet
+}
+
+// This is not a full implementation of OIDC
+// It *just* checks the token against the keys of an extra provider.
+func NewJWTFromExtraProviderAuthenticator(
+	headerName, issuer, issuerName, providerURL string) Authenticator {
+	return &JWTFromExtraProviderAuthenticator{
+		headerName:   headerName,
+		issuer:       issuer,
+		isserName:    issuerName,
+		remoteKeySet: goidc.NewRemoteKeySet(context.Background(), providerURL+"/keys"),
+	}
+}
+
+func (s *JWTFromExtraProviderAuthenticator) Authenticate(w http.ResponseWriter, r *http.Request) (*common.User, bool, error) {
+	logger := common.RequestLogger(r, "JWT extra token authenticator")
+
+	if len(r.Header[s.headerName]) != 1 {
+		return nil, false, nil
+	}
+
+	rawIDToken := strings.TrimPrefix(r.Header[s.headerName][0], bearerPrefix)
+
+	// Validate token
+	verifier := goidc.NewVerifier(s.issuer, s.remoteKeySet, &goidc.Config{SkipClientIDCheck: true})
+	jwt, err := verifier.Verify(r.Context(), rawIDToken)
+	if err != nil {
+		return nil, false, &common.AuthenticatorSpecificError{Err: err}
+	}
+	// Retrieve the USERID_CLAIM and the GROUPS_CLAIM
+	var claims map[string]interface{}
+	if claimErr := jwt.Claims(&claims); claimErr != nil {
+		logger.Errorf("Retrieving user claims failed: %v", claimErr)
+		return nil, false, &common.AuthenticatorSpecificError{Err: claimErr}
+	}
+	userID, groups, claimErr := s.retrieveUserIDGroupsClaims(claims)
+	if claimErr != nil {
+		return nil, false, &common.AuthenticatorSpecificError{Err: claimErr}
+	}
+
+	user := common.User{
+		Name:   userID + ":" + s.isserName,
```

Comment:
I find it strange we're mutating the user-id here to include issuer information

Reply:
I agree. I was not sure if want to do it or not.
I added it because I find the information good to have. We can look at the proxy logs and just by checking the logs we know that the token used to authenticate the request is from the "cv-infra" and not the "local" cluster:
![image](https://github.com/aristanetworks/oidc-authservice/assets/100247/cb144776-cdfe-4b76-ab6c-c95ab59c5a3e)

Also there is a "USERID_TRANSFORMERS" which I find weird as well, so this is kind-of doing something similar.
If you don't like it, we can remove it, we could try to send another header as well to log this information in the logs, it's more work, maybe we can do this in a future iteration.